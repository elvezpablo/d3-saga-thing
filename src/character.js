const character = () => {
  console.log("Egiller");
};

export { character };
