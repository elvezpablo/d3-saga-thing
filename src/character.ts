import { select } from "d3";

const RADIUS = 10;

const character = () => {
  console.log("Egiller");
};

const draw = () => {
  const data = [
    { "x_axis": 30, "y_axis": 30, "radius": 20, "color": "green" },
    { "x_axis": 70, "y_axis": 70, "radius": 20, "color": "purple" },
    { "x_axis": 110, "y_axis": 100, "radius": 20, "color": "red" }
  ]
  const svg = select("#sketch");
  const circles = svg.selectAll("circle")
    .data(data)
    .enter()
    .append("circle");

  circles
    .attr("cx", function (d) { return d.x_axis; })
    .attr("cy", function (d) { return d.y_axis; })
    .attr("r", RADIUS)
    .style("fill", function (d) { return d.color; });
}

draw();

export { character };
