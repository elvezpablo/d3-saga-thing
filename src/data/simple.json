{
    "name": "GP",
    "children": [
        {
            "name": "PA",
            "children": [
                {
                    "name": "CA1"
                },
                {
                    "name": "CA2"
                },
                {
                    "name": "CA3",
                    "children": [
                        {
                            "name": "BCA3"
                        }
                    ]
                }
            ]
        },
        {
            "name": "PB",
            "children": []
        },
        {
            "name": "PC",
            "children": [
                {
                    "name": "CC1"
                },
                {
                    "name": "CC2"
                }
            ]
        }
    ]
}