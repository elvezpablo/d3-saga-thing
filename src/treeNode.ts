interface ITreeNode {
  x: number;
  y: number;
  finalY: number;
  mod: number;
  parent?: ITreeNode;
  prevSibling?: ITreeNode;
  children?: ITreeNode[];
  data?: any | undefined;
  collapse: boolean;
}

class TreeNode implements ITreeNode {
  x = 0;
  y = 0;
  finalY = 0;
  modifier = 0;
  parent?: ITreeNode;
  prevSibling?: ITreeNode;
  mod = 0;
  children?: ITreeNode[] = [];
  data?: Object;
  collapse = false;

  constructor(x: number, y: number, parent?: ITreeNode, prevSibling?: ITreeNode, data?: any) {
    this.x = x;
    this.y = y;

    this.parent = parent;
    this.prevSibling = prevSibling;

    this.data = data.name;
  }
}



export { ITreeNode, TreeNode };
