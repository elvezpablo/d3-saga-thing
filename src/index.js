import { character } from "./character";

const func = () => {
  console.log("whew");
  character();
};

func();

console.log("dud");
