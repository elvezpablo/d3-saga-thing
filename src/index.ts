import { character } from "./character";
import simple from "../__tests__/data/simple.json";
// import { html } from "d3";

const func = () => {
  console.log("awe yeassddh!");
  character();
  console.log(simple)
};

func();

console.log("dud");
