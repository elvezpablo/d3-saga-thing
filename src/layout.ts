import { TreeNode, ITreeNode } from "./treeNode";

enum SPACING {
  LEFT_PADDING = 3,
}

// https://www.digitalocean.com/community/tutorials/js-tree-traversal
// https://github.com/zchtodd/recurser/blob/master/static/recurser.js

function initialY(node: ITreeNode) {
  const { children, y, prevSibling } = node;
  const { LEFT_PADDING } = SPACING;

  for (let i = 0; i < children.length; i++) {
    initialY(children[i]);
  }

  node.y = prevSibling ? prevSibling.y + LEFT_PADDING : 0;

  if (children.length == 1) {
    node.mod = y;
  } else if (children.length >= 2) {
    let minY = Infinity;
    let maxY = -minY;
    for (let i = 0; i < children.length; i++) {
      minY = Math.min(minY, children[i].y);
      maxY = Math.max(maxY, children[i].y);
    }
    node.mod = y - (maxY - minY) / 2;
  }
}

function calculateFinalValues(node, modSum) {
  node.finalY = node.y + modSum;
  for (let i = 0; i < node.children.length; i++) {
    calculateFinalValues(node.children[i], node.modifier + modSum);
  }
}

function getContour(root, val, func) {
  let nodes = [root];
  while (nodes.length) {
    let node = nodes.shift();
    nodes = nodes.concat(node.children);
    val = func(val, node.finalY);
  }
  return val;
}

function shiftDown(root, shiftValue) {
  let nodes = [root];
  while (nodes.length) {
    let node = nodes.shift();
    nodes = nodes.concat(node.children);
    node.finalY += shiftValue;
  }
}

function fixNodeConflicts(root) {
  for (let i = 0; i < root.children.length; i++) {
    fixNodeConflicts(root.children[i]);
  }

  for (let i = 0; i < root.children.length - 1; i++) {
    // Get the bottom-most contour position of the current node
    let botContour = getContour(root.children[i], -Infinity, Math.max);

    // Get the topmost contour position of the node underneath the current one
    let topContour = getContour(root.children[i + 1], Infinity, Math.min);

    if (botContour >= topContour) {
      shiftDown(root.children[i + 1], botContour - topContour + 3);
    }
  }
}

function buildTree(dataNode, parent, prevSibling, level) {
  let root = new Node(level, 0, parent, prevSibling, dataNode);
  for (let i = 0; i < dataNode.children.length; i++) {
    root.children.push(
      buildTree(
        dataNode.children[i],
        root,
        i >= 1 ? root.children[i - 1] : null,
        level + 1
      )
    );
  }
  return root;
}

function updateYVals(root) {
  let minYVal = Infinity;
  let nodes = [root];
  while (nodes.length) {
    let node = nodes.shift();
    nodes = nodes.concat(node.children);
    if (node.finalY < minYVal) {
      minYVal = node.finalY;
    }
  }

  nodes = [root];
  while (nodes.length) {
    let node = nodes.shift();
    nodes = nodes.concat(node.children);
    node.finalY += Math.abs(minYVal);
  }
}

function getDimensions(root) {
  let minWidth = Infinity;
  let maxWidth = -minWidth;

  let minHeight = Infinity;
  let maxHeight = -minWidth;

  let nodes = [root];
  while (nodes.length) {
    let node = nodes.shift();
    nodes = nodes.concat(node.children);

    if (node.x < minWidth) {
      minWidth = node.x;
    }

    if (node.x > maxWidth) {
      maxWidth = node.x;
    }

    if (node.finalY < minHeight) {
      minHeight = node.finalY;
    }

    if (node.finalY > maxHeight) {
      maxHeight = node.finalY;
    }
  }
  return [maxWidth - minWidth, maxHeight - minHeight];
}
