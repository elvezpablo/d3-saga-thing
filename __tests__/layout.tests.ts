import { ITreeNode, TreeNode } from "../src/treeNode";
import {
  buildTree,
  calculateInitialValues,
  calculateFinalValues,
  updateYVals,
  fixNodeConflicts,
  getContour,
  shiftDown
} from "../src/layout"

import simple from "./data/simple.json";

// need to traverse in post order
test("buildTree", () => {
  let root = buildTree(simple, 0, null, null);
  // console.log(root.children[0]);  
  expect(root.children.length).toBe(3);
  expect(root.children[0].data.name).toBe("parentA")
  expect(root.children[0].children[0].data.name).toBe("childA1")
  expect(root.children[2].children[0].data.name).toBe("childC1")
});

test("calculateInitialValues", () => {
  let root = buildTree(simple, 0, null, null);
  calculateInitialValues(root);
  expect(root.y).toBe(0);
  expect(root).toMatchSnapshot();
});

test("calculateFinalValues", () => {
  let root = buildTree(simple, 0, null, null);
  calculateFinalValues(root, 0);
  expect(root).toMatchSnapshot();
});

test("getContour", () => {
  const root = new TreeNode(0, 0, null, null, { name: "root" });
  root.finalY = 20;
  const prevSib = new TreeNode(0, 0, root, null, { name: "prev" });
  prevSib.finalY = 10;
  root.children.push(prevSib);
  const nextSib = new TreeNode(0, 0, root, prevSib, { name: "next" });
  nextSib.finalY = 20;
  root.children.push(nextSib);
  const lastSib = new TreeNode(0, 0, root, nextSib, { name: "last" })
  lastSib.finalY = 30;
  root.children.push(lastSib);

  // Get the bottom-most contour position of the current node
  const bottom = getContour(root, Number.MIN_VALUE, Math.max);
  expect(bottom).toBe(30);
  const top = getContour(root, Number.MAX_VALUE, Math.min);
  expect(top).toBe(10);
  root.finalY = 5;
  const newTop = getContour(root, Number.MAX_VALUE, Math.min);

  expect(newTop).toBe(5);
})

test("fixNodeConflicts", () => {
  let root = buildTree(simple, 0, null, null);
  calculateInitialValues(root);
  calculateFinalValues(root, 0);
  updateYVals(root);
  fixNodeConflicts(root);
  expect(root).toMatchSnapshot();
});

test("shiftDown", () => {
  const root = new TreeNode(0, 0, null, null, { name: "root" });
  root.finalY = 20;
  const prevSib = new TreeNode(0, 0, root, null, { name: "prev" });
  prevSib.finalY = 10;
  root.children.push(prevSib);
  const nextSib = new TreeNode(0, 0, root, prevSib, { name: "next" });
  nextSib.finalY = 20;
  root.children.push(nextSib);
  const lastSib = new TreeNode(0, 0, root, nextSib, { name: "last" })
  lastSib.finalY = 30;
  root.children.push(lastSib);

  shiftDown(root, 10);

  expect(root.finalY).toBe(30);
  expect(prevSib.finalY).toBe(20);
  expect(nextSib.finalY).toBe(30);

})