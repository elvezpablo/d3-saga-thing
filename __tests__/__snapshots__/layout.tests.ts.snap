// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateInitialValues 1`] = `
TreeNode {
  "children": Array [
    TreeNode {
      "children": Array [
        TreeNode {
          "children": Array [],
          "collapse": false,
          "data": Object {
            "name": "childA1",
          },
          "finalY": 0,
          "mod": 0,
          "modifier": 0,
          "parent": [Circular],
          "prevSibling": null,
          "x": 0,
          "y": 0,
        },
        TreeNode {
          "children": Array [],
          "collapse": false,
          "data": Object {
            "name": "childA2",
          },
          "finalY": 0,
          "mod": 0,
          "modifier": 0,
          "parent": [Circular],
          "prevSibling": TreeNode {
            "children": Array [],
            "collapse": false,
            "data": Object {
              "name": "childA1",
            },
            "finalY": 0,
            "mod": 0,
            "modifier": 0,
            "parent": [Circular],
            "prevSibling": null,
            "x": 0,
            "y": 0,
          },
          "x": 1,
          "y": 3,
        },
        TreeNode {
          "children": Array [],
          "collapse": false,
          "data": Object {
            "name": "childA3",
          },
          "finalY": 0,
          "mod": 0,
          "modifier": 0,
          "parent": [Circular],
          "prevSibling": TreeNode {
            "children": Array [],
            "collapse": false,
            "data": Object {
              "name": "childA2",
            },
            "finalY": 0,
            "mod": 0,
            "modifier": 0,
            "parent": [Circular],
            "prevSibling": TreeNode {
              "children": Array [],
              "collapse": false,
              "data": Object {
                "name": "childA1",
              },
              "finalY": 0,
              "mod": 0,
              "modifier": 0,
              "parent": [Circular],
              "prevSibling": null,
              "x": 0,
              "y": 0,
            },
            "x": 1,
            "y": 3,
          },
          "x": 2,
          "y": 6,
        },
      ],
      "collapse": false,
      "data": Object {
        "children": Array [
          Object {
            "name": "childA1",
          },
          Object {
            "name": "childA2",
          },
          Object {
            "name": "childA3",
          },
        ],
        "name": "parentA",
      },
      "finalY": 0,
      "mod": -3,
      "modifier": 0,
      "parent": [Circular],
      "prevSibling": null,
      "x": 0,
      "y": 0,
    },
    TreeNode {
      "children": Array [],
      "collapse": false,
      "data": Object {
        "children": Array [],
        "name": "parentB",
      },
      "finalY": 0,
      "mod": 0,
      "modifier": 0,
      "parent": [Circular],
      "prevSibling": TreeNode {
        "children": Array [
          TreeNode {
            "children": Array [],
            "collapse": false,
            "data": Object {
              "name": "childA1",
            },
            "finalY": 0,
            "mod": 0,
            "modifier": 0,
            "parent": [Circular],
            "prevSibling": null,
            "x": 0,
            "y": 0,
          },
          TreeNode {
            "children": Array [],
            "collapse": false,
            "data": Object {
              "name": "childA2",
            },
            "finalY": 0,
            "mod": 0,
            "modifier": 0,
            "parent": [Circular],
            "prevSibling": TreeNode {
              "children": Array [],
              "collapse": false,
              "data": Object {
                "name": "childA1",
              },
              "finalY": 0,
              "mod": 0,
              "modifier": 0,
              "parent": [Circular],
              "prevSibling": null,
              "x": 0,
              "y": 0,
            },
            "x": 1,
            "y": 3,
          },
          TreeNode {
            "children": Array [],
            "collapse": false,
            "data": Object {
              "name": "childA3",
            },
            "finalY": 0,
            "mod": 0,
            "modifier": 0,
            "parent": [Circular],
            "prevSibling": TreeNode {
              "children": Array [],
              "collapse": false,
              "data": Object {
                "name": "childA2",
              },
              "finalY": 0,
              "mod": 0,
              "modifier": 0,
              "parent": [Circular],
              "prevSibling": TreeNode {
                "children": Array [],
                "collapse": false,
                "data": Object {
                  "name": "childA1",
                },
                "finalY": 0,
                "mod": 0,
                "modifier": 0,
                "parent": [Circular],
                "prevSibling": null,
                "x": 0,
                "y": 0,
              },
              "x": 1,
              "y": 3,
            },
            "x": 2,
            "y": 6,
          },
        ],
        "collapse": false,
        "data": Object {
          "children": Array [
            Object {
              "name": "childA1",
            },
            Object {
              "name": "childA2",
            },
            Object {
              "name": "childA3",
            },
          ],
          "name": "parentA",
        },
        "finalY": 0,
        "mod": -3,
        "modifier": 0,
        "parent": [Circular],
        "prevSibling": null,
        "x": 0,
        "y": 0,
      },
      "x": 1,
      "y": 3,
    },
    TreeNode {
      "children": Array [
        TreeNode {
          "children": Array [],
          "collapse": false,
          "data": Object {
            "name": "childC1",
          },
          "finalY": 0,
          "mod": 0,
          "modifier": 0,
          "parent": [Circular],
          "prevSibling": null,
          "x": 2,
          "y": 0,
        },
        TreeNode {
          "children": Array [],
          "collapse": false,
          "data": Object {
            "name": "childC2",
          },
          "finalY": 0,
          "mod": 0,
          "modifier": 0,
          "parent": [Circular],
          "prevSibling": TreeNode {
            "children": Array [],
            "collapse": false,
            "data": Object {
              "name": "childC1",
            },
            "finalY": 0,
            "mod": 0,
            "modifier": 0,
            "parent": [Circular],
            "prevSibling": null,
            "x": 2,
            "y": 0,
          },
          "x": 3,
          "y": 3,
        },
      ],
      "collapse": false,
      "data": Object {
        "children": Array [
          Object {
            "name": "childC1",
          },
          Object {
            "name": "childC2",
          },
        ],
        "name": "parentC",
      },
      "finalY": 0,
      "mod": -1.5,
      "modifier": 0,
      "parent": [Circular],
      "prevSibling": TreeNode {
        "children": Array [],
        "collapse": false,
        "data": Object {
          "children": Array [],
          "name": "parentB",
        },
        "finalY": 0,
        "mod": 0,
        "modifier": 0,
        "parent": [Circular],
        "prevSibling": TreeNode {
          "children": Array [
            TreeNode {
              "children": Array [],
              "collapse": false,
              "data": Object {
                "name": "childA1",
              },
              "finalY": 0,
              "mod": 0,
              "modifier": 0,
              "parent": [Circular],
              "prevSibling": null,
              "x": 0,
              "y": 0,
            },
            TreeNode {
              "children": Array [],
              "collapse": false,
              "data": Object {
                "name": "childA2",
              },
              "finalY": 0,
              "mod": 0,
              "modifier": 0,
              "parent": [Circular],
              "prevSibling": TreeNode {
                "children": Array [],
                "collapse": false,
                "data": Object {
                  "name": "childA1",
                },
                "finalY": 0,
                "mod": 0,
                "modifier": 0,
                "parent": [Circular],
                "prevSibling": null,
                "x": 0,
                "y": 0,
              },
              "x": 1,
              "y": 3,
            },
            TreeNode {
              "children": Array [],
              "collapse": false,
              "data": Object {
                "name": "childA3",
              },
              "finalY": 0,
              "mod": 0,
              "modifier": 0,
              "parent": [Circular],
              "prevSibling": TreeNode {
                "children": Array [],
                "collapse": false,
                "data": Object {
                  "name": "childA2",
                },
                "finalY": 0,
                "mod": 0,
                "modifier": 0,
                "parent": [Circular],
                "prevSibling": TreeNode {
                  "children": Array [],
                  "collapse": false,
                  "data": Object {
                    "name": "childA1",
                  },
                  "finalY": 0,
                  "mod": 0,
                  "modifier": 0,
                  "parent": [Circular],
                  "prevSibling": null,
                  "x": 0,
                  "y": 0,
                },
                "x": 1,
                "y": 3,
              },
              "x": 2,
              "y": 6,
            },
          ],
          "collapse": false,
          "data": Object {
            "children": Array [
              Object {
                "name": "childA1",
              },
              Object {
                "name": "childA2",
              },
              Object {
                "name": "childA3",
              },
            ],
            "name": "parentA",
          },
          "finalY": 0,
          "mod": -3,
          "modifier": 0,
          "parent": [Circular],
          "prevSibling": null,
          "x": 0,
          "y": 0,
        },
        "x": 1,
        "y": 3,
      },
      "x": 2,
      "y": 6,
    },
  ],
  "collapse": false,
  "data": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "name": "childA1",
          },
          Object {
            "name": "childA2",
          },
          Object {
            "name": "childA3",
          },
        ],
        "name": "parentA",
      },
      Object {
        "children": Array [],
        "name": "parentB",
      },
      Object {
        "children": Array [
          Object {
            "name": "childC1",
          },
          Object {
            "name": "childC2",
          },
        ],
        "name": "parentC",
      },
    ],
    "name": "grandparent",
  },
  "finalY": 0,
  "mod": -3,
  "modifier": 0,
  "parent": null,
  "prevSibling": null,
  "x": 0,
  "y": 0,
}
`;
